web/packages/design/src/utils/useResizeObserver.ts

```diff
@@ -16,46 +16,78 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import { RefObject, useCallback, useLayoutEffect } from 'react';
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  type RefCallback,
+} from 'react';
 
 /**
- * useResizeObserver sets up a ResizeObserver for ref and calls callback on each resize.
+ * useResizeObserver creates a ResizeObserver which fires a callback
+ * when the provided element is resized. The callback is called with the
+ * ResizeObserverEntry. The observer is disconnected when the element is
+ * unmounted.
  *
- * It does not fire if ref.current.contentRect.height is zero, to account for a special case in
- * Connect where tabs are hidden using `display: none;`.
+ * Returns a ref callback to attach to the element to be observed – the element
+ * must be a HTMLElement but may be conditionally null.
  *
- * Uses a layout effect underneath. If ref is conditionally rendered, set enabled to false when ref
- * is null.
+ * @example
+ * // Basic usage
+ * const Component = () => {
+ *   const handleResize = (entry: ResizeObserverEntry) => {
+ *     console.log({ entryContentRect: entry.contentRect });
+ *   };
+ *
+ *   const ref = useResizeObserver<HTMLDivElement>(handleResize);
+ *
+ *   return <div ref={ref}>This div is being observed</div>;
+ * };
  */
-export function useResizeObserver(
-  ref: RefObject<HTMLElement>,
+export const useResizeObserver = <T extends HTMLElement = HTMLElement>(
   callback: (entry: ResizeObserverEntry) => void,
-  { enabled = true }
-): void {
-  const effect = useCallback(() => {
-    if (!ref.current || !enabled) {
-      return;
-    }
+  { fireOnZeroHeight = true } = {}
+): RefCallback<T> => {
+  const callbackRef = useRef(callback);
 
-    const observer = new ResizeObserver(entries => {
-      const entry = entries[0];
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
 
-      // In Connect, when a tab becomes active, its outermost DOM element switches from `display:
-      // none` to `display: flex`. This callback is then fired with the height reported as zero.
-      // To avoid unnecessary calls to callback, return early here.
-      if (entry.contentRect.height === 0) {
-        return;
-      }
+  const observer = useMemo(
+    () =>
+      new ResizeObserver(([entry]) => {
+        if (!entry || (!fireOnZeroHeight && entry.contentRect.height === 0))
+          return;
+        callbackRef.current?.(entry);
+      }),
+    [fireOnZeroHeight]
+  );
 
-      callback(entry);
-    });
+  const elementRef = useRef<T | null>(null);
 
-    observer.observe(ref.current);
+  useEffect(() => {
+    const element = elementRef.current;
+    if (element) {
+      observer.observe(element);
```

Comment:
Is this necessary? From a quick test, this makes the hook call `observer.observe` twice on the same element after the first render.

I recall [that advice about keeping an effect and its cleanup symmetrical](https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount), but this doesn't let us keep them symmetrical anyway.